### PR TITLE
Resolve PR comments in #92

### PR DESCRIPTION
**Pull request by GDRDABARERA**

### CODE REVIEW

File: components/org.wso2.carbon.identity.mgt.store.connector.jdbc/src/main/java/org/wso2/carbon/identity/mgt/store/connector/jdbc/connector/JDBCIdentityStoreConnector.java [hunk 1]
```diff
@@ -527,20 +526,33 @@ public IdentityStoreConnectorConfig getIdentityStoreConfig() {
...
+            } else {
+                throw new IdentityStoreConnectorException("There are no attributes to find users.");
+            }
             if (log.isDebugEnabled()) {
                 log.debug("{} sql queries loaded for database type: {}.", sqlQueries.size(), databaseType);
             }
         } else {
-            throw new StoreException("Invalid or unsupported database type specified in the configuration.");
+            throw new IdentityStoreConnectorException("Invalid or unsupported database type specified in the" +
+                    " configuration.");
```
Comment (kasunbg) [inline components/org.wso2.carbon.identity.mgt.store.connector.jdbc/src/main/java/org/wso2/carbon/identity/mgt/store/connector/jdbc/connector/JDBCIdentityStoreConnector.java]: Error messages need more context than this. Please mention the `databaseType` that the user has set.

File: components/org.wso2.carbon.identity.mgt.store.connector.jdbc/src/main/java/org/wso2/carbon/identity/mgt/store/connector/jdbc/queries/MySQLFamilySQLQueryFactory.java [hunk 2]
```diff
@@ -285,16 +285,16 @@ public MySQLFamilySQLQueryFactory() {
...
-                            " WHERE ATTR_ID = (SELECT ID FROM UM_ATTRIBUTES WHERE ATTR_NAME = '")
-                    .append(attribute.getAttributeName())
-                    .append("' ) AND ATTR_VALUE = '")
-                    .append(attribute.getAttributeValue())
-                    .append("')");
+                            " WHERE ATTR_ID = (SELECT ID FROM UM_ATTRIBUTES WHERE ATTR_NAME = ")
+                    .append(":attri").append((count * 2) - 1)
+                    .append("; ) AND ATTR_VALUE = ")
+                    .append(":attri").append(count * 2)
+                    .append("; )");
```
Comment (kasunbg) [inline components/org.wso2.carbon.identity.mgt.store.connector.jdbc/src/main/java/org/wso2/carbon/identity/mgt/store/connector/jdbc/queries/MySQLFamilySQLQueryFactory.java]: @GDRDABARERA I think this code is still vulnerable to SQL injection. You should use PreparedStatement instead to construct the SQL query. Is it not possible here? cc @Pushpalanka

File: components/org.wso2.carbon.identity.mgt.store.connector.jdbc/src/main/java/org/wso2/carbon/identity/mgt/store/connector/jdbc/connector/JDBCIdentityStoreConnector.java [hunk 3]
```diff
@@ -527,20 +526,33 @@ public IdentityStoreConnectorConfig getIdentityStoreConfig() {
...
                     unitOfWork.getConnection(), sqlQuerryForUserAttributes);
+            int count = 1;
+            for (Attribute attribute : attributes) {
+                if (count <= attributes.size()) {
+                    getUsersNamedPreparedStatement.setString(ConnectorConstants.SQLPlaceholders.ATTRIBUTE
+                            .concat(Integer.toString((count * 2) - 1)), attribute.getAttributeName());
+                    getUsersNamedPreparedStatement.setString(ConnectorConstants.SQLPlaceholders.ATTRIBUTE
+                            .concat(Integer.toString(count * 2)), attribute.getAttributeValue());
+                    ++count;
+                }
```
Comment (kasunbg) [inline components/org.wso2.carbon.identity.mgt.store.connector.jdbc/src/main/java/org/wso2/carbon/identity/mgt/store/connector/jdbc/connector/JDBCIdentityStoreConnector.java]: Which PR comment are you addressing here?

File: components/org.wso2.carbon.identity.mgt.store.connector.jdbc/src/main/java/org/wso2/carbon/identity/mgt/store/connector/jdbc/queries/MySQLFamilySQLQueryFactory.java (hunk 2 above)
Comment (GDRDABARERA) [reply components/org.wso2.carbon.identity.mgt.store.connector.jdbc/src/main/java/org/wso2/carbon/identity/mgt/store/connector/jdbc/queries/MySQLFamilySQLQueryFactory.java]: The issue is fixed now. The confusion might be resolved now

File: components/org.wso2.carbon.identity.mgt.store.connector.jdbc/src/main/java/org/wso2/carbon/identity/mgt/store/connector/jdbc/queries/MySQLFamilySQLQueryFactory.java [hunk 4]
```diff
@@ -306,4 +310,42 @@ public String getQuerryForUserIdFromMultipleAttributes(List<Attribute> attribute
...
+            if (count < attributes.size()) {
+                getUniqueUserQuerry.append(" AND ");
+            }
+            ++count;
+        }
+        getUniqueUserQuerry.append(" GROUP BY USER_UNIQUE_ID LIMIT ");
+        getUniqueUserQuerry.append(length);
+        getUniqueUserQuerry.append(";");
+
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.mgt.store.connector.jdbc/src/main/java/org/wso2/carbon/identity/mgt/store/connector/jdbc/queries/MySQLFamilySQLQueryFactory.java]: Please remove unwanted empty lines

File: components/org.wso2.carbon.identity.mgt.store.connector.jdbc/src/main/java/org/wso2/carbon/identity/mgt/store/connector/jdbc/queries/MySQLFamilySQLQueryFactory.java [hunk 5]
```diff
@@ -306,4 +310,42 @@ public String getQuerryForUserIdFromMultipleAttributes(List<Attribute> attribute
 
         return getUniqueUserQuerry.toString();
     }
+
+    public PreparedStatement getPreparedStatementFromMultipleAttributes(
+            List<Attribute> attributes, int offset, int length, UnitOfWork unitOfWork) throws SQLException {
+        StringBuilder getUniqueUserQuerry = new StringBuilder();
+        getUniqueUserQuerry.append("SELECT UM_USER.USER_UNIQUE_ID FROM UM_USER WHERE UM_USER.ID IN");
+        int count = 1;
+        while (count <= attributes.size()) {
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.mgt.store.connector.jdbc/src/main/java/org/wso2/carbon/identity/mgt/store/connector/jdbc/queries/MySQLFamilySQLQueryFactory.java]: What will happen if the list of attributes is empty, i.e. size == 0 ?

File: components/org.wso2.carbon.identity.mgt.store.connector.jdbc/src/main/java/org/wso2/carbon/identity/mgt/store/connector/jdbc/queries/MySQLFamilySQLQueryFactory.java [hunk 6]
```diff
@@ -306,4 +310,42 @@ public String getQuerryForUserIdFromMultipleAttributes(List<Attribute> attribute
...
+                    .append(":").append(ConnectorConstants.SQLPlaceholders.ATTRIBUTE).append(count * 2)
+                    .append("; )");
+            if (count < attributes.size()) {
+                getUniqueUserQuerry.append(" AND ");
+            }
+            ++count;
+        }
+        getUniqueUserQuerry.append(" GROUP BY USER_UNIQUE_ID LIMIT ");
+        getUniqueUserQuerry.append(length);
+        getUniqueUserQuerry.append(";");
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.mgt.store.connector.jdbc/src/main/java/org/wso2/carbon/identity/mgt/store/connector/jdbc/queries/MySQLFamilySQLQueryFactory.java]: why ";" ?

File: components/org.wso2.carbon.identity.mgt.store.connector.jdbc/src/main/java/org/wso2/carbon/identity/mgt/store/connector/jdbc/queries/MySQLFamilySQLQueryFactory.java (hunk 5 above)
Comment (GDRDABARERA) [reply components/org.wso2.carbon.identity.mgt.store.connector.jdbc/src/main/java/org/wso2/carbon/identity/mgt/store/connector/jdbc/queries/MySQLFamilySQLQueryFactory.java]: This is handled in the method which calls this method, Only if attributes.size()>0, the prepared statement is generated else throw new IdentityStoreConnectorException which tells that There are no attributes to find users. 
getUsers(List<Attribute> attributes, int offset, int length) --> JDBCIdentityStoreConnector.java

Comment (ruwanta) [reply components/org.wso2.carbon.identity.mgt.store.connector.jdbc/src/main/java/org/wso2/carbon/identity/mgt/store/connector/jdbc/queries/MySQLFamilySQLQueryFactory.java]: This is a public method. There is no control over who calls this method.
Any method should give valid result on any range of input parameters or throw a checked exception if parameters are our of range. We can not assume the other party will call with correct parameters.

**Pull request closed without merging**